engine/src/test/scala/cromwell/engine/workflow/WorkflowDockerLookupActorSpec.scala

```diff
@@ -0,0 +1,391 @@
+package cromwell.engine.workflow
+
+import akka.actor.{ActorRef, Props}
+import akka.testkit.{ImplicitSender, TestActorRef, TestProbe}
+import com.typesafe.config.ConfigFactory
+import cromwell.core.actor.StreamIntegration.BackPressure
+import cromwell.core.{TestKitSuite, WorkflowId}
+import cromwell.database.slick.SlickDatabase
+import cromwell.database.sql.tables.DockerHashStoreEntry
+import cromwell.docker.DockerHashActor.{DockerHashFailedResponse, DockerHashSuccessResponse}
+import cromwell.docker.{DockerHashRequest, DockerHashResult, DockerImageIdentifier, DockerImageIdentifierWithoutHash}
+import cromwell.engine.workflow.WorkflowActor.{RestartExistingWorkflow, StartMode, StartNewWorkflow}
+import cromwell.engine.workflow.WorkflowDockerLookupActor.{DockerHashActorTimeout, WorkflowDockerLookupFailure, WorkflowDockerTerminalFailure}
+import cromwell.engine.workflow.WorkflowDockerLookupActorSpec.TestWorkflowDockerLookupActor
+import cromwell.services.ServicesStore._
+import cromwell.services.SingletonServicesStore
+import org.scalatest.{BeforeAndAfter, FlatSpecLike, Matchers}
+import org.specs2.mock.Mockito
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+
+
+class WorkflowDockerLookupActorSpec extends TestKitSuite("WorkflowDockerLookupActorSpecSystem") with FlatSpecLike with Matchers with ImplicitSender with BeforeAndAfter with Mockito {
+  it should "wait and resubmit the docker request when it gets a backpressure message" in {
+    val dockerValue = "ubuntu:latest"
+    val dockerId = DockerImageIdentifier.fromString(dockerValue).get.asInstanceOf[DockerImageIdentifierWithoutHash]
+
+    val workflowId = WorkflowId.randomId()
+    val dockerHashingActor = TestProbe()
+    val backpressureWaitTime = 2 seconds
+
+    val lookupActor = TestActorRef(Props(new TestWorkflowDockerLookupActor(workflowId, dockerHashingActor.ref, StartNewWorkflow, backpressureWaitTime)), self)
+    val request = DockerHashRequest(dockerId)
+    lookupActor ! request
+
+    dockerHashingActor.expectMsg(request)
+    dockerHashingActor.reply(BackPressure(request))
+    // Give a couple of seconds of margin to account for test latency etc...
+    dockerHashingActor.expectMsg(backpressureWaitTime.+(2 seconds), request)
+  }
+
+  it should "not look up the same tag again after a successful lookup" in {
+    val dockerValue = "ubuntu:latest"
+    val dockerId = DockerImageIdentifier.fromString(dockerValue).get.asInstanceOf[DockerImageIdentifierWithoutHash]
+
+    val workflowId = WorkflowId.randomId()
+    val dockerHashingActor = TestProbe()
+
+    val databaseConfig = ConfigFactory.load.getConfig("database")
+
+    var numWrites = 0
+
+    val databaseInterface = new SlickDatabase(databaseConfig) {
```

Comment:
The implementation is different for each test, but I'll see if any DRYing is possible.